Source/SwiftLintFramework/Rules/MultilineFunctionChainsRule.swift

```diff
@@ -0,0 +1,221 @@
+                let offset = callDotOffset(file: file, callRange: range),
+                let line = file.contents.bridge().lineAndCharacter(forCharacterOffset: offset)?.line else {
+                    return nil
+            }
+            return (dotLine: line, dotOffset: offset, range: range)
+        }
+
+        let uniqueLines = calls
+            .map { $0.dotLine }
+            .reduce(into: Set<Int>()) { (result: inout Set<Int>, line: Int) in
+                result.insert(line)
+            }
+
+        if uniqueLines.count == 1 { return [] }
+
+        // The first call (last here) is allowed to not have a leading newline.
+        let noLeadingNewlineViolations = calls
+            .dropLast()
+            .filter { line in
+                !callHasLeadingNewline(file: file, callRange: line.range)
+            }
+
+        return noLeadingNewlineViolations.map { $0.dotOffset }
+    }
+
+    private static let whitespaceDotRegex = regex("\\s*\\.")
+
+    private func callDotOffset(file: File, callRange: NSRange) -> Int? {
+        guard
+            let range = file.contents.bridge().byteRangeToNSRange(start: callRange.location, length: callRange.length),
+            case let regex = type(of: self).whitespaceDotRegex,
+            let match = regex.matches(in: file.contents, options: [], range: range).last?.range else {
+                return nil
+        }
+        return match.location + match.length - 1
+    }
+
+    private static let newlineWhitespaceDotRegex = regex("\\n\\s*\\.")
+
+    private func callHasLeadingNewline(file: File, callRange: NSRange) -> Bool {
+        guard
+            let range = file.contents.bridge().byteRangeToNSRange(start: callRange.location, length: callRange.length),
+            case let regex = type(of: self).newlineWhitespaceDotRegex,
+            regex.firstMatch(in: file.contents, options: [], range: range) != nil else {
+                return false
+        }
+        return true
+    }
+
+    private func callRanges(file: File,
+                            kind: SwiftExpressionKind,
+                            dictionary: [String: SourceKitRepresentable],
+                            parentCallName: String? = nil) -> [NSRange] {
+        guard
+            kind == .call,
+            case let contents = file.contents.bridge(),
+            let offset = dictionary.nameOffset,
+            let length = dictionary.nameLength,
+            let name = contents.substringWithByteRange(start: offset, length: length) else {
+                return []
+        }
+
+        let subcalls = dictionary.subcalls
+
+        if subcalls.isEmpty, let parentCallName = parentCallName, parentCallName.starts(with: name) {
+            return [NSRange(location: offset, length: length)]
+        }
+
+        return subcalls.flatMap { call -> [NSRange] in
+            guard let range = callRange(file: file, call: call, parentName: name, parentNameOffset: offset) else {
+                return []
```

Comment:
shouldn't we return `callRanges(file: file, kind: .call, dictionary: call, parentCallName: name)` here?

Reply:
In this case, since there's no chained subcall we don't need to descend further on, so we can just bail out and return early. I'll go ahead and add a comment to make it clear 👍